File: Happy/Services/VoiceService.swift
import Foundation
import AVFoundation
import Combine

// MARK: - Voice Configuration

/// Voice configuration for the ElevenLabs integration
enum VoiceConfig {
    /// Development agent ID
    static let agentIdDev = "agent_7801k2c0r5hjfraa1kdbytpvs6yt"

    /// Production agent ID
    static let agentIdProd = "agent_6701k211syvvegba4kt7m68nxjmw"

    /// Get the appropriate agent ID based on build configuration
    static var agentId: String {
        #if DEBUG
        return agentIdDev
        #else
        return agentIdProd
        #endif
    }

    /// ElevenLabs API base URL
    static let apiBaseURL = "https://api.elevenlabs.io/v1"

    /// Default TTS model
    static let ttsModel = "eleven_turbo_v2"

    /// Default voice ID (Rachel - conversational)
    static let defaultVoiceId = "EXAVITQu4vr4xnSDxMaL"

    /// Enable debug logging
    static var enableDebugLogging: Bool {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }
}

// MARK: - Voice Status

/// Voice connection status
enum VoiceStatus: String {
    case disconnected
    case connecting
    case connected
    case error
}

/// Voice conversation mode
enum VoiceMode: String {
    case speaking
    case listening
    case idle
}

// MARK: - Voice State

/// Voice session state
struct VoiceState {
    var status: VoiceStatus = .disconnected
    var mode: VoiceMode = .idle
    var isMuted: Bool = false
    var activeSessionId: String? = nil
    var conversationId: String? = nil
    var error: String? = nil
    var voiceLanguage: String = "en"
}

// MARK: - Voice Session Config

/// Configuration for starting a voice session
struct VoiceSessionConfig {
    let sessionId: String
    var initialContext: String?
    var token: String?
    var agentId: String?
}

// MARK: - Voice Service

/// Voice service for macOS using ElevenLabs Conversational AI
///
/// Uses AVFoundation for audio playback and URLSession for API communication.
/// Provides real-time voice communication with Claude Code sessions.
///
/// Example usage:
/// ```swift
/// let voiceService = VoiceService.shared
///
/// // Start a voice session
/// await voiceService.startSession(config: VoiceSessionConfig(
///     sessionId: "session-123",
///     initialContext: "User is viewing a coding session"
/// ))
///
/// // Send a contextual update
/// voiceService.sendContextualUpdate("User clicked on file.ts")
///
/// // End the session
/// await voiceService.endSession()
/// ```
@MainActor
class VoiceService: NSObject, ObservableObject {

    // MARK: - Singleton

    static let shared = VoiceService()

    // MARK: - Published State

    @Published private(set) var state = VoiceState()

    // MARK: - Private Properties

    private var audioPlayer: AVAudioPlayer?
    private var audioEngine: AVAudioEngine?
    private var webSocketTask: URLSessionWebSocketTask?
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Computed Properties

    /// Whether a voice session is currently active
    var isActive: Bool {
        state.status == .connected
    }

    /// Whether currently connecting
    var isConnecting: Bool {
        state.status == .connecting
    }

    /// Whether there's an error
    var hasError: Bool {
        state.status == .error && state.error != nil
    }

    /// Human-readable status message
    var statusMessage: String {
        switch state.status {
        case .disconnected:
            return "Voice assistant offline"
        case .connecting:
            return "Connecting..."
        case .connected:
            if state.isMuted {
                return "Muted"
            }
            switch state.mode {
            case .speaking:
                return "Speaking..."
            case .listening:
                return "Listening..."
            case .idle:
                return "Ready"
            }
        case .error:
            return state.error ?? "Connection error"
        }
    }

    // MARK: - Initialization

    private override init() {
        super.init()
        setupAudioEngine()
    }

    // MARK: - Audio Setup

    private func setupAudioEngine() {
        audioEngine = AVAudioEngine()

        // Configure audio session for voice
        // Note: macOS uses a different API than iOS for audio sessions
    }

    // MARK: - Session Management

    /// Start a voice session
    /// - Parameter config: Configuration for the voice session
    func startSession(config: VoiceSessionConfig) async {
        // Request microphone permission
        guard await requestMicrophonePermission() else {
            updateState { $0.status = .error; $0.error = "Microphone permission denied" }
            return
        }

        updateState {
            $0.status = .connecting
            $0.activeSessionId = config.sessionId
        }

        let agentId = config.agentId ?? VoiceConfig.agentId

        if VoiceConfig.enableDebugLogging {
            print("[Voice] Starting session with agent: \(agentId)")
            print("[Voice] Language: \(state.voiceLanguage)")
        }

        do {
            try await connectToVoiceService(config: config, agentId: agentId)

            updateState {
                $0.status = .connected
                $0.mode = .listening
            }

            if VoiceConfig.enableDebugLogging {
                print("[Voice] Session started successfully")
            }
        } catch {
            print("[Voice] Failed to start session: \(error)")
            updateState {
                $0.status = .error
                $0.error = error.localizedDescription
            }
        }
    }

    /// Connect to the voice service
    private func connectToVoiceService(config: VoiceSessionConfig, agentId: String) async throws {
        // Implementation would:
        // 1. Get signed URL from ElevenLabs
        // 2. Establish WebSocket connection
        // 3. Set up audio capture and playback

        if VoiceConfig.enableDebugLogging {
            print("[Voice] Would connect to voice service with:", [
                "agentId": agentId,
                "sessionId": config.sessionId,
                "language": state.voiceLanguage,
                "hasToken": config.token != nil
            ])
        }

        // For now, simulate successful connection
        // Real implementation would use URLSession WebSocket
    }

    /// End the current voice session
    func endSession() async {
        do {
            // Close WebSocket connection
            webSocketTask?.cancel(with: .normalClosure, reason: nil)
            webSocketTask = nil

            // Stop audio engine
            audioEngine?.stop()
            audioPlayer?.stop()
            audioPlayer = nil

            updateState {
                $0.status = .disconnected
                $0.mode = .idle
                $0.activeSessionId = nil
                $0.conversationId = nil
            }

            if VoiceConfig.enableDebugLogging {
                print("[Voice] Session ended")
            }
        }
    }

    // MARK: - Messaging

    /// Send a text message to the voice assistant
    /// - Parameter message: The message to send
    func sendTextMessage(_ message: String) {
        guard state.status == .connected else {
            print("[Voice] No active session")
            return
        }

        if VoiceConfig.enableDebugLogging {
            print("[Voice] Sending text message: \(message)")
        }

        // Would send via WebSocket
    }

    /// Send a contextual update to the voice assistant
    /// - Parameter update: The contextual update to send
    func sendContextualUpdate(_ update: String) {
        guard state.status == .connected else {
            if VoiceConfig.enableDebugLogging {
                print("[Voice] No active session, skipping context update")
            }
            return
        }

        if VoiceConfig.enableDebugLogging {
            print("[Voice] Sending contextual update: \(update)")
        }

        // Would send via WebSocket
    }

    // MARK: - Controls

    /// Toggle mute state
    func toggleMute() {
        updateState { $0.isMuted.toggle() }

        // Would mute/unmute microphone input
    }

    /// Set mute state
    /// - Parameter muted: Whether to mute
    func setMuted(_ muted: Bool) {
        updateState { $0.isMuted = muted }

        // Would mute/unmute microphone input
    }

    /// Set voice language preference
    /// - Parameter language: The language code (e.g., "en", "es")
    func setVoiceLanguage(_ language: String) {
        updateState { $0.voiceLanguage = language }
    }

    // MARK: - Permissions

    /// Request microphone permission
    /// - Returns: Whether permission was granted
    private func requestMicrophonePermission() async -> Bool {
        // On macOS, microphone permission is requested when accessing the input
        // Check current authorization status
        switch AVCaptureDevice.authorizationStatus(for: .audio) {
        case .authorized:
            return true
        case .notDetermined:
            return await withCheckedContinuation { continuation in
                AVCaptureDevice.requestAccess(for: .audio) { granted in
                    continuation.resume(returning: granted)
                }
            }
        case .denied, .restricted:
            return false
        @unknown default:
            return false
        }
    }

    // MARK: - State Updates

    /// Update state with a closure
    private func updateState(_ update: (inout VoiceState) -> Void) {
        var newState = state
        update(&newState)
        state = newState
    }
}

// MARK: - Voice Language Mapping

/// Map user preference codes to ElevenLabs language codes
func getElevenLabsLanguageCode(_ preference: String) -> String {
    let languageMap: [String: String] = [
        "en": "en",
        "es": "es",
        "ru": "ru",
        "pl": "pl",
        "pt": "pt",
        "ca": "ca",
        "zh-Hans": "zh"
    ]
    return languageMap[preference] ?? "en"
}

File: apps/mobile/src/services/voice/VoiceService.ts
/**
 * Voice Service for NativeScript
 *
 * Mobile voice service implementation using native audio APIs.
 * Since ElevenLabs doesn't have a NativeScript SDK, this implementation
 * uses the REST API for TTS and native audio players for playback.
 *
 * Platform-specific audio handling:
 * - iOS: Uses AVAudioPlayer via NativeScript
 * - Android: Uses MediaPlayer via NativeScript
 *
 * @example
 * ```typescript
 * import { voiceService } from '@/services/voice';
 *
 * await voiceService.startSession({
 *   sessionId: 'session-123',
 *   initialContext: 'User is viewing a coding session'
 * });
 *
 * await voiceService.endSession();
 * ```
 */

import { Application, isIOS, isAndroid } from '@nativescript/core';
import type { VoiceSession, VoiceSessionConfig, VoiceStatus, VoiceMode } from './types';
import { ELEVENLABS_CONFIG, VOICE_CONFIG, getElevenLabsLanguageCode } from './config';

/**
 * Voice state management
 */
interface VoiceState {
    status: VoiceStatus;
    mode: VoiceMode;
    isMuted: boolean;
    activeSessionId: string | null;
    conversationId: string | null;
    error: string | null;
    voiceLanguage: string;
}

/**
 * Global voice state
 */
const voiceState: VoiceState = {
    status: 'disconnected',
    mode: 'idle',
    isMuted: false,
    activeSessionId: null,
    conversationId: null,
    error: null,
    voiceLanguage: 'en',
};

/**
 * State change listeners
 */
type StateListener = (state: VoiceState) => void;
const stateListeners: StateListener[] = [];

/**
 * Notify all listeners of state change
 */
function notifyStateChange(): void {
    for (const listener of stateListeners) {
        listener({ ...voiceState });
    }
}

/**
 * Update state and notify listeners
 */
function updateState(updates: Partial<VoiceState>): void {
    Object.assign(voiceState, updates);
    notifyStateChange();
}

/**
 * Request microphone permission on mobile
 */
async function requestMicrophonePermission(): Promise<boolean> {
    try {
        if (isIOS) {
            // iOS: Request microphone permission via AVAudioSession
            // Note: Actual implementation would use proper iOS audio session APIs
            // const AudioSession = (NSObject as any).extend({});
            // Simplified - returns true for now
            return true;
        } else if (isAndroid) {
            // Android: Request RECORD_AUDIO permission
            const permissions = (android as any).Manifest.permission;
            const context = Application.android.context;
            const hasPermission = (androidx as any).core.content.ContextCompat.checkSelfPermission(
                context,
                permissions.RECORD_AUDIO
            ) === (android as any).content.pm.PackageManager.PERMISSION_GRANTED;

            if (!hasPermission) {
                // Would need to request permission through activity
                return false;
            }
            return true;
        }
        return false;
    } catch (error) {
        console.error('[Voice] Failed to request microphone permission:', error);
        return false;
    }
}

/**
 * Voice Service Implementation for NativeScript
 *
 * Uses REST API for ElevenLabs TTS since there's no native SDK.
 * For real-time conversation, uses WebSocket connection where available.
 */
class VoiceServiceImpl implements VoiceSession {
    /**
     * Start a voice session
     */
    async startSession(config: VoiceSessionConfig): Promise<void> {
        // Request microphone permission first
        const hasPermission = await requestMicrophonePermission();
        if (!hasPermission) {
            updateState({ status: 'error', error: 'Microphone permission denied' });
            return;
        }

        updateState({
            status: 'connecting',
            activeSessionId: config.sessionId,
        });

        try {
            const language = getElevenLabsLanguageCode(voiceState.voiceLanguage);
            const agentId = config.agentId ?? ELEVENLABS_CONFIG.agentId;

            if (VOICE_CONFIG.ENABLE_DEBUG_LOGGING) {
                console.log('[Voice] Starting mobile session with agent:', agentId);
                console.log('[Voice] Language:', language);
            }

            // Note: Full implementation would establish WebSocket connection
            // to ElevenLabs Conversational AI endpoint for real-time voice.
            // For now, we set up the structure for TTS-based voice.

            await this.connectToVoiceService(config, agentId, language);

            updateState({
                status: 'connected',
                mode: 'listening',
            });

            if (VOICE_CONFIG.ENABLE_DEBUG_LOGGING) {
                console.log('[Voice] Mobile session started successfully');
            }
        } catch (error) {
            console.error('[Voice] Failed to start session:', error);
            updateState({
                status: 'error',
                error: error instanceof Error ? error.message : 'Failed to start voice session',
            });
        }
    }

    /**
     * Connect to voice service
     */
    private async connectToVoiceService(
        config: VoiceSessionConfig,
        agentId: string,
        language: string
    ): Promise<void> {
        // Implementation would:
        // 1. Establish WebSocket to ElevenLabs
        // 2. Set up audio capture from microphone
        // 3. Set up audio playback for responses

        if (VOICE_CONFIG.ENABLE_DEBUG_LOGGING) {
            console.log('[Voice] Would connect to voice service with:', {
                agentId,
                sessionId: config.sessionId,
                language,
                platform: isIOS ? 'iOS' : 'Android',
            });
        }
    }

    /**
     * End the current voice session
     */
    async endSession(): Promise<void> {
        try {
            // Stop audio capture and playback
            // Close WebSocket connection

            updateState({
                status: 'disconnected',
                mode: 'idle',
                activeSessionId: null,
                conversationId: null,
            });

            if (VOICE_CONFIG.ENABLE_DEBUG_LOGGING) {
                console.log('[Voice] Session ended');
            }
        } catch (error) {
            console.error('[Voice] Failed to end session:', error);
        }
    }

    /**
     * Send a text message to the voice assistant
     */
    sendTextMessage(message: string): void {
        if (voiceState.status !== 'connected') {
            console.warn('[Voice] No active session');
            return;
        }

        if (VOICE_CONFIG.ENABLE_DEBUG_LOGGING) {
            console.log('[Voice] Sending text message:', message);
        }

        // Would send via WebSocket to ElevenLabs
    }

    /**
     * Send a contextual update to the voice assistant
     */
    sendContextualUpdate(update: string): void {
        if (voiceState.status !== 'connected') {
            if (VOICE_CONFIG.ENABLE_DEBUG_LOGGING) {
                console.log('[Voice] No active session, skipping context update');
            }
            return;
        }

        if (VOICE_CONFIG.ENABLE_DEBUG_LOGGING) {
            console.log('[Voice] Sending contextual update:', update);
        }

        // Would send via WebSocket to ElevenLabs
    }

    /**
     * Check if a session is currently active
     */
    isSessionActive(): boolean {
        return voiceState.status === 'connected';
    }

    /**
     * Get current voice state
     */
    getState(): VoiceState {
        return { ...voiceState };
    }

    /**
     * Subscribe to state changes
     */
    subscribe(listener: StateListener): () => void {
        stateListeners.push(listener);
        return () => {
            const index = stateListeners.indexOf(listener);
            if (index > -1) {
                stateListeners.splice(index, 1);
            }
        };
    }

    /**
     * Toggle mute state
     */
    toggleMute(): void {
        updateState({ isMuted: !voiceState.isMuted });
    }

    /**
     * Set mute state
     */
    setMuted(muted: boolean): void {
        updateState({ isMuted: muted });
    }

    /**
     * Set voice language preference
     */
    setVoiceLanguage(language: string): void {
        updateState({ voiceLanguage: language });
    }
}

/**
 * Singleton voice service instance
 */
export const voiceService = new VoiceServiceImpl();

/**
 * Start a voice session for a given session ID
 */
export async function startVoiceSession(
    sessionId: string,
    initialContext?: string
): Promise<void> {
    await voiceService.startSession({ sessionId, initialContext });
}

/**
 * End the current voice session
 */
export async function endVoiceSession(): Promise<void> {
    await voiceService.endSession();
}

/**
 * Check if voice session is active
 */
export function isVoiceSessionActive(): boolean {
    return voiceService.isSessionActive();
}

/**
 * Get current voice state
 */
export function getVoiceState(): VoiceState {
    return voiceService.getState();
}
